src/hotspot/share/gc/z/zBarrierSetNMethod.cpp

```diff
@@ -105,3 +105,33 @@ oop ZBarrierSetNMethod::oop_load_no_keepalive(const nmethod* nm, int index) {
 oop ZBarrierSetNMethod::oop_load_phantom(const nmethod* nm, int index) {
   return ZNMethod::oop_load_phantom(nm, index);
 }
+
+void ZBarrierSetNMethod::arm_with(nmethod* nm, int value) {
+  assert((value & not_entrant) == 0, "not_entrant bit is reserved");
+  ZLocker<ZReentrantLock> locker(ZNMethod::lock_for_nmethod(nm));
+  // Preserve the sticky bit
+  if (is_not_entrant(nm)) {
+    value |= not_entrant;
+  }
+  set_guard_value(nm, value);
```

Comment:
We can avoid redundant patching since we have the lock: Only update if the value is not already there. This happens quite often.